### ISSUE TITLE
From ViT models to audio

### ISSUE THREAD

**Issue by Antoine101** [kkoutini/PaSST#45]
Hi Khaled,

In your code, there is the possibility to create a ViT architecture and load the corresponding pretrained weights (like "vit_tiny_patch16_224").

Do we agree that such architectures only work with similar size inputs (224*224 for example)? If so, how did you finetune a model on Audioset that was initially trained on Imagenet (going from 224*224 to 128*998 for example)? Is this procedure in some code in your repo?

I read the AST paper I guess you took inspiration from and they talk about it in some details.
I was just wondering how I would do the whole process (ImageNet -> AudioSet -> ESC50) on my end.

Thanks a lot.

Antoine

**Comment by kkoutini** [comment]
Hi Antoine,

Yes, the code should support more architectures.

If the input channels are different, the input channels are averaged [here](https://github.com/kkoutini/PaSST/blob/4519e4605989b8c2e62dccb5b928af9bf7bf8602/models/helpers/vit_helpers.py#L107) and [here](https://github.com/kkoutini/PaSST/blob/4519e4605989b8c2e62dccb5b928af9bf7bf8602/models/helpers/vit_helpers.py#L40)

If the input size is different (for example, `224x224` to `128x998`), the only thing that is changed is the positional embeddings, this is done [here](https://github.com/kkoutini/PaSST/blob/2a5c818afcc2a215b2a1aaf1ed8be71f89d43201/models/passt.py#L656)
in short, the positional embeddings are interpolated to match the new size (similar to AST). After that, they are averged over time/freq  to produce freq/time positional embeddings.

**Comment by Antoine101** [comment]
Great, I'll have a look at all that!
Thanks a lot.

**Comment by Antoine101** [comment]
Regarding that question, I see in the code that there are different lists of architectures available between `get_model` function, `default_cfg` dictionnary and architecture functions.

`default_cfg` seems to be the most exhaustive but not every architecture in this dict is covered in `get_model` or has a dedicated function that calls `_create_vision_transformer`.
Is it just because you didn't test them all or didn't have time to implement everything or is there another specific reason?

See below:
![image](https://github.com/kkoutini/PaSST/assets/48209504/9559f7eb-8c1f-4a91-b1c3-72c19ff71b9b)

![image](https://github.com/kkoutini/PaSST/assets/48209504/bd096782-7347-4a7e-8a17-9b8423ee85eb)

![image](https://github.com/kkoutini/PaSST/assets/48209504/70dc74d2-873e-4516-862f-d3e50087cabf)

Thanks a lot.

**Comment by kkoutini** [comment]
Hi,
I got the basis code from tim library. where it has links for different models then I added the models that I trained one by one in the same fashion with a link to download the weights. The missing ones are the ones that I didn't use. However, I believe it should work if you add more ViT in the same way.

**Comment by Antoine101** [comment]
Hi Khaled, 

Thanks for the answer.

Regarding your first reply on this thread, concerning the adaptation/averaging of input channels, why is there a sum on dim=1 in the code instead of a mean? In adapt_input_conv [here](https://github.com/kkoutini/PaSST/blob/4519e4605989b8c2e62dccb5b928af9bf7bf8602/models/helpers/vit_helpers.py#L40)

**Comment by kkoutini** [comment]
I think you are right `mean` should work better

**Comment by Antoine101** [comment]
Great, thanks for the confirmation!
